FlowCrypt/Functionality/Mail Provider/MessageOperations Provider/Imap+MessageOperations.swift

```diff
@@ -37,117 +61,106 @@ extension Imap: MessageOperationsProvider {
-                .start(self.finalizeAsVoid("moveMsg", resolve, reject, retry: { self.moveMsg(with: identifier, folder: folder, destFolder: destFolder) }))
+                .copyMessagesOperation(withFolder: folder, uids: MCOIndexSet(index: UInt64(id)), destFolder: destFolder)
+                .start { (error, _)in
+                    if let error = error {
+                        continuation.resume(throwing: ImapError.providerError(error))
+                    } else {
+                        continuation.resume(returning: ())
+                    }
+                }
         }
     }
 
-    // MARK: - delete
-    func delete(message: Message, form folderPath: String?) -> Promise<Void> {
-        Promise<Void> { [weak self] _, reject in
-            guard let self = self else { return reject(AppErr.nilSelf) }
+    func delete(message: Message, form folderPath: String?) async throws {
+        guard let identifier = message.identifier.intId else {
+            throw ImapError.missedMessageInfo("intId")
+        }
 
-            guard let identifier = message.identifier.intId else {
-                return reject(ImapError.missedMessageInfo("intId"))
-            }
+        guard let folderPath = folderPath else {
+            throw ImapError.missedMessageInfo("folderPath")
+        }
 
-            guard let folderPath = folderPath else {
-                return reject(ImapError.missedMessageInfo("folderPath"))
-            }
+        try await pushUpdatedMsgFlags(with: identifier, folder: folderPath, flags: MCOMessageFlag.deleted)
+        try await expungeMsgs(folder: folderPath)
 
-            try awaitPromise(self.pushUpdatedMsgFlags(with: identifier, folder: folderPath, flags: MCOMessageFlag.deleted))
-            try awaitPromise(self.expungeMsgs(folder: folderPath))
-        }
     }
 
-    private func pushUpdatedMsgFlags(with identifier: Int, folder: String, flags: MCOMessageFlag) -> Promise<Void> {
-        Promise { [weak self] resolve, reject in
-            guard let self = self else { return reject(AppErr.nilSelf) }
+    private func pushUpdatedMsgFlags(with identifier: Int, folder: String, flags: MCOMessageFlag) async throws {
+        try await withCheckedThrowingContinuation { [weak self] (continuation: CheckedContinuation<Void, Error>) in
+            guard let self = self else {
+                return continuation.resume(throwing: AppErr.nilSelf)
+            }
 
-            let retry = self.pushUpdatedMsgFlags(with: identifier, folder: folder, flags: flags)
             self.imapSess?
                 .storeFlagsOperation(
                     withFolder: folder,
                     uids: MCOIndexSet(index: UInt64(identifier)),
                     kind: MCOIMAPStoreFlagsRequestKind.set,
                     flags: flags
                 )
-                .start(self.finalizeVoid("updateMsgFlags", resolve, reject, retry: { retry }))
```

Comment:
These finalize methods fulfill an important role - logging, renewing access tokens, retrying but only if the error is appropriate to retry & after renewing the access token. Has this been re-implemented for async/await?

Reply:
I think no, not sure this could be done with async/await.
As for me I think we are misusing this concept

Reply:
If you remove or significantly change functionality from the app, don't do it silently, there needs to be a discussion about that.

Logging is significant, and so is access token renewal.

I don't know if Gmail API will do this automatically or not, maybe yes. But token renewal on IMAP/SMTP will break without this.

Of course retry and access token renewal mechanism can be implemented in async/await, and it becomes easier, too.

Since IMAP/SMTP is not our primary target now (main customer, and most users, use Gmail API), I'll get this breakage merged into master, but then I'll have to fix it. I'll file an issue.